### PR TITLE
Update of prelude-go to replace gocode backend with lsp-mode and gopls

### PR DESCRIPTION
**Pull request by hut8**
Addresses https://github.com/bbatsov/prelude/issues/1361

My goals are:

- Remove gocode as much as possible.
- go-projectile ends up installing it (and requires eldoc-go) which is not ideal, but other minor modes (like eldoc-go) require it. My development experience is fine without it, but it would be nice to have as long as it stops requiring gocode. I opened https://github.com/dougm/go-projectile/issues/23 for this.
- Remove go-eldoc because of hard dependency on gocode
- Remove company-go because of hard dependency on gocode, and company-lsp provides better functionality.
- Add lsp-mode and lsp-ui and integrate them with go-mode (as long as gopls is in $PATH, it works flawlessly).
- Add docs

### CODE REVIEW

**Comment by bbatsov** [comment]
I'm fine with the spirit of the proposed changes. I don't even know what `gocode` is and whether it has any advantages for LSP. :-)

File: modules/prelude-go.el [hunk 1]
```diff
@@ -61,14 +64,18 @@
     ;; gofmt on save
     (add-hook 'before-save-hook 'gofmt-before-save nil t)
 
-    ;; stop whitespace being highlighted
-    (whitespace-toggle-options '(tabs))
+    (setq company-idle-delay 0)
```
Comment (bbatsov) [inline modules/prelude-go.el:67]: Seems weird to have mode-specific `company-mode` configuration. Any special reason for this?

File: modules/prelude-go.el [hunk 2]
```diff
@@ -31,10 +31,13 @@
...
 
 (prelude-require-packages '(go-mode
-                            company-go
-                            go-eldoc
-                            go-projectile
-                            gotest))
+                            lsp-mode
+                            lsp-ui
+                            company
+                            yasnippet
```
Comment (bbatsov) [inline modules/prelude-go.el:37]: Any special reason for adding `yas` just for go?

**Comment by bbatsov** [comment]
Btw, it'd be good if you also updated the related documentation for Go programming with Prelude (or create it if it's missing).

File: modules/prelude-go.el (hunk 1 above)
Comment (hut8) [reply modules/prelude-go.el]: The special reason is that I forgot to remove this 😄

File: modules/prelude-go.el (hunk 2 above)
Comment (hut8) [reply modules/prelude-go.el]: Removed. I'll add it in the documentation which I'm about to write; `go` is highly redundant which makes yas very useful but this belongs in a doc, not automatically loaded.

**Comment by hut8** [comment]
@bbatsov docs added as per your request. Feedback welcome 😄

**Comment by bbatsov** [comment]
@hut8 Looks good to me.

**Comment by hut8** [comment]
I think this should be good now; ready for final review.

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**